org.osgi.test.assertj.framework/src/main/java/org/osgi/test/assertj/servicereference/AbstractServiceReferenceAssert.java

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright (c) OSGi Alliance (2020). All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.osgi.test.assertj.servicereference;
+
+import static org.osgi.test.assertj.bundle.BundleAssert.BUNDLE;
+
+import java.util.Arrays;
+import java.util.List;
+
+import org.assertj.core.api.AbstractComparableAssert;
+import org.assertj.core.api.InstanceOfAssertFactories;
+import org.assertj.core.api.InstanceOfAssertFactory;
+import org.assertj.core.api.ListAssert;
+import org.osgi.framework.Bundle;
+import org.osgi.framework.FrameworkUtil;
+import org.osgi.framework.ServiceReference;
+import org.osgi.test.assertj.bundle.BundleAssert;
+import org.osgi.test.assertj.dictionary.DictionaryAssert;
+import org.osgi.test.common.dictionary.Dictionaries;
+
+public abstract class AbstractServiceReferenceAssert<SELF extends AbstractServiceReferenceAssert<SELF, ACTUAL>, ACTUAL extends ServiceReference<?>>
```

Comment:
I am not sure why this is necessary. Are you testing framework implementation to see if they properly implement the OSGi spec? Otherwise, why not just assume the framework is properly implemented. The support we have OSGi testing should be focused on testing bundles people write. To the extent they are useful for testing OSGi framework specification implementation, that is great, but we should not worry too much about that.

Reply:
As described above the main goal it to have a fluent was to came to DictionaryAssert.
The other Methods are wishes from @kriegfrj  https://github.com/osgi/osgi-test/pull/245#discussion_r540613956

Reply:
But if we update to osgi.core:7..0.0, then you just need to do a dictionary assert on ServiceReference.getProperties and then you don't need this class.

Reply:
> is think that a ServiceReferenceAssert is very usefull for a lot of developers.
> will add more tests like `hasObjectClass`

I think you are just testing the framework impl here. Do you think a bundle developer needs to test that his service is properly registered?

Reply:
Sure, you do not need this for a single with DS- registered Component. 
And I will never use more then the bridge to 
```
 `DictionaryAssert` via `hasServicePropertiesThat`
 `BundleAssert` via `hasBundleThat`
``` 
But @kriegfrj  asked me to implement all test when i touch this `ServiceReferenceAssert`.

I had a look on the Tests of DataInMotion and talked to them. They used such tests and checks change-counts and tests services in this detail. And they use this in there OSGi-Spec implementations AND normal bundles. This happened most times when observing ServiceEvents ( like witha mecanism like https://github.com/osgi/osgi-test/pull/257) and then asserting the Reference of a cached Event.

Reply:
> But if we update to osgi.core:7..0.0, then you just need to do a dictionary assert on ServiceReference.getProperties and then you don't need this class.

sure. you can to it by your own. I just like the fluent way.

Reply:
I think this is very useful. We e.g. have a Component that can receive a configuration a few other components  and registers a couple service. When we test it, we check if the services are around, have the right properties and that they are bound to each other correctly, because modify e.g. the reference targets via the configuration.

Reply:
> I am not sure why this is necessary. Are you testing framework implementation to see if they properly implement the OSGi spec? Otherwise, why not just assume the framework is properly implemented. The support we have OSGi testing should be focused on testing bundles people write. To the extent they are useful for testing OSGi framework specification implementation, that is great, but we should not worry too much about that.

I sympathise with this, which is why I hadn't gone to the effort of writing `ServiceReferenceAssert` myself - I thought it would be rarely used (mainly in framework compliance testing scenarios, which is useful though not our top priority). And if you're using DS, then you won't often touch a `ServiceReference` directly in your code anyway. That said, there are still times when you need to directly access `ServiceReference` and I can see that this assertion might be useful in those cases too.

Add to that, if someone else has identified a need in their own code and was keen to do the work, I thought "why not let them"?

The only concern I have that, if it is to be done, it ought to be done as properly and completely as is practical. It's typically not too much effort to implement most of the useful assertion methods for an assert class once you've already committed to the new assertion class itself. And it can be annoying to think that you have the assertion available only to find it's missing the method you actually needed.